redhat-6.2-with-source/redhat-6.2/en_6.2/misc/src/anaconda/isys/modutils/man/depmod.man
.\" Copyright (c) 1994, 1995 Jacques Gelinas (jacques@solucorp.qc.ca)
.\" Copyright (c) 1995 Bjorn Ekwall (bj0rn@blox.se)
.\" This program is distributed according to the Gnu General Public License.
.\" See the file COPYING in the kernel source directory /linux
.\" $Id: depmod.1,v 1.1.1.1 1998/01/06 20:51:07 ewt Exp $
.\"
.TH DEPMOD 1 "March 17, 1996" Linux "Linux Module Support"
.SH NAME
depmod, modprobe \- handle loadable modules automatically
.SH SYNOPSIS
.B depmod
[ \-a ]
.br
.B depmod
[ \-a version ]
.br
.B depmod
module1.o module2.o ...
.PP
.B modprobe
module.o [symbol=value ...]
.br
.B modprobe
\-t tag pattern
.br
.B modprobe
\-a \-t tag pattern
.br
.B modprobe
\-l [ \-t tag ] pattern
.br
.B modprobe
\-r module
.br
.B modprobe
\-c
.br
.SH DESCRIPTION
These utilities are intended
to make a Linux modular kernel manageable for all users,
administrators and distribution maintainers.
.PP
.B Depmod
creates a "Makefile"-like dependency file, based on the symbols it finds
in the set of modules mentioned on the command line (or in a default place).
This dependency file can later be used by
.B modprobe
to automatically load the relevant module(s).
.PP
.B Modprobe
is used to load a set of modules, either a single module,
a stack of dependant modules,
or all modules that are marked with a specified tag.
.PP
.B Modprobe
will automatically load all base modules needed in a module stack,
as described by the dependency file modules.dep.
If the loading of one of these modules fails, the whole current stack
of modules will be unloaded (by rmmod) automatically.
.PP
.B Modprobe
has two ways of loading modules. One way (the probe mode) will
try to load a module out of a list (defined by
.B pattern
).
It stops loading as soon as one module load successfully.
This can be used to autoload one ethernet driver out of a list for example.
The other way, is to load all modules from a list.
This can be used to load some modules at boot time.
.PP
With the option
.B -r,
modprobe will automatically unload a stack of modules,
similar to the way
.B rmmod -r
does.
.PP
Option -l combined with option -t list all available modules
of a certain type.
An enhanced
.B mount
command could use the command

	modprobe -l -t fs

to get the list of all file system drivers available and on
request load the proper one.
So, the mount command could become more generic as well...
(The kerneld solve this without changing the mount utility)
.PP
Option -c will print all configuration (default + configuration file).
.PP
The normal use of
.B depmod
is to include the line "/sbin/depmod -a" in one of the rc-files
in /etc/rc.d, so that the correct module dependencies will be available
immediately after booting the system.
.br
Note that it is also possible to create the dependency file immediately
after compiling a new kernel.
If you do "depmod -a 1.3.99" when you have compiled kernel 1.3.99 and
its modules the first time, while still running e.g. 1.3.98, the file will
be created in the correct place.
.br
Note however that the dependencies on the kernel will not be guaranteed
to be correct in this case!
.PP
Option
.B -d
put depmod in debug mode. It outputs all command it is issuing.
.Pp
Option
.B -e
output the list of unresolved symbol for each module, Normally depmod
only output the list of unloadable modules.
.Pp
Option
.B -v
output the list of all processed modules.
.Pp
Modules may be located at different place in the filesystem,
but there will always be some need to override this,
especially for module developers.
We expect some official standard will emerge, defined by the FSSTND.
Until that time you might as well use this suggested directory structure.
.SH CONFIGURATION
The behaviour of
.B depmod
and
.B modprobe
can be adjusted by the (optional) configuration file
.B /etc/conf.modules
.PP
The configuration file consists of a set of lines.
.br
All empty lines, and all text on a line after a '#', will be ignored.
.br
Lines may be continued by ending the line with a '\\'.
.br
The remaining lines should all conform to one of the following formats:

  keep
  parameter=value
  options module symbol=value ...
  alias module real_name
  pre-install module command ...
  install module command ...
  post-install module command ...
  pre-remove module command ...
  remove module command ...
  post-remove module command ...

All values in the "parameter" lines will be processed by a shell,
which means that "shell tricks" like wild-cards
and commands enclosed in back-quotes can be used:

  path[misc]=/lib/modules/1.1.5?
  path[net]=/lib/modules/`uname -r`

Parameters may be repeated multiple times.
.PP
These are the legal parameters:
.TP
keep
If this word is found on a line
.B before
any lines that contain the
.B path
descriptions, the default set of paths will be saved, and thus added to.
Otherwise the normal behaviour is that the default set will be
.B replaced
by the set of paths in the configuration file.
.TP
depfile=DEPFILE_PATH
This is the path to the dependency file that will be created by
.B depmod
and used by
.B modprobe.
.TP
path=SOME_PATH
The
.B path
parameter specifies a directory to search for the modules.
.TP
path[tag]=SOME_PATH
The path parameter can carry an optional tag.
This tells us a little more about the purpose of the modules in
this directory and allows some automated operations by
.B modprobe.
The tag is appended to the "path" keyword enclose in square brackets.
If the tag is missing, the tag "misc" is assumed.
.br
One very useful tag is
.B boot,
which can be used to mark all modules that should be loaded at boot-time.
.PP
If the configuration file '/etc/conf.modules' is missing,
or if any parameter is not overridden,
the following defaults are assumed:
.PP
  depfile=/lib/modules/`uname -r`/modules.dep
  path[boot]=/lib/modules

  path[fs]=/lib/modules/`uname -r`
  path[misc]=/lib/modules/`uname -r`
  path[net]=/lib/modules/`uname -r`
  path[scsi]=/lib/modules/`uname -r`

  path[fs]=/lib/modules/default
  path[misc]=/lib/modules/default
  path[net]=/lib/modules/default
  path[scsi]=/lib/modules/default

  path[fs]=/lib/modules
  path[misc]=/lib/modules
  path[net]=/lib/modules
  path[scsi]=/lib/modules
.PP
All "option" lines specify the default options that are needed for
a module, as in:

  modprobe de620 bnc=1

These options will be overridden by any options given on the
.B modprobe
command line.
.br
It is possible to have an "option" line for aliased module names as well
as for the non-aliased name.
This is useful for e.g. the dummy module:

  alias dummy0 dummy
  options dummy0 -o dummy0

.PP
The "alias" lines can be used to give alias names to modules.
A line in /etc/conf.modules that looks like this:

  alias iso9660 isofs

makes it possible to write
.B "modprobe iso9660"
although there is no such module available.
.br
Note that the line:

  alias some_module off

will make modprobe ignore requests to load that module.
This is usually used in conjunction with
.B kerneld.
.TP
Commands
The configuration lines
  pre-install module command ...
  install module command ...
  post-install module command ...
  pre-remove module command ...
  remove module command ...
  post-remove module command ...
.br
can be used when one wants some specific commands to be executed when
a module is inserted or removed.
All text after the module name will be interpreted as the command text.
.br
Note that the pre- and post-remove commands will \fBnot\fR be executed
if a module is "autocleaned" by kerneld!
Look for the up-coming support for persistent module storage instead.
.SH STRATEGY
The idea is that
.B modprobe
will look first at the directory containing
modules compiled for the current release of the kernel.
If the module is not found there,
.B modprobe
will look in the directory containing modules for a default release.
.PP
When you install a new linux, the modules should be moved to a directory
related to the release (and version) of the kernel you are installing.
Then you should do a symlink from this directory to the "default" directory.
.PP
Each time you compile a new kernel, the command
.B "make modules_install"
will create a new directory, but won't change the default.
.PP
When you get a module unrelated to the kernel distribution
you should place it in one of the version-independent directories
under /lib/modules.
.PP
This is the default strategy, which can be overridden in /etc/conf.modules.
.SH EXAMPLES
.TP
modprobe -t net
Load one of the modules that are stored in the directory tagged "net".
Each module are tried until one succeed (default: /lib/modules/net).
.TP
modprobe -a -t boot
All modules that are stored in the directory tagged "boot" will
be loaded (default: /lib/modules/boot).
.TP
modprobe slip.o
This will attempt to load the module slhc.o if it was not previously loaded,
since the slip module needs the functionality in the slhc module.
This dependency will be described in the file "modules.dep" that was
created automatically by
.B depmod
.TP
modprobe -r slip.o
will unload slip.o. It will also unload slhc.o automatically,
unless it is used by some other module as well (like e.g. ppp.o).
.SH FILES
.nf
/etc/conf.modules, (and /etc/modules.conf)
/lib/modules/*/modules.dep,
/lib/modules/*
.fi
.SH SEE ALSO
lsmod(1),
kerneld(8),
ksyms(1),
modules(2),
.SH REQUIERED UTILITIES
insmod(1),
nm(1)
rmmod(1),
.SH NOTES
The pattern supplied to modprobe will often be escaped to ensure
that it is evaluated in the proper context

.SH AUTHOR
Jacques Gelinas (jack@solucorp.qc.ca)
.br
Bjorn Ekwall (bj0rn@blox.se)
.SH BUGS
Naah...


modutils/ChangeLog
Thu Sep 11 08:43:03 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* Release version 2.1.55.

Wed Sep 10 16:34:41 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* genksyms/genksyms.c (crc_prefix): New variable.
	(export_symbol): Use it.
	(main): New option -p sets it.
	* man/genksyms.8: Update documentation.
	Idea from Jacques Gelinas <jack@solucorp.qc.ca>.

Wed Sep 10 16:09:37 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* insmod/Makefile: Allow insmod & rmmod to run as a combined binary.
	* insmod/main.c: New file.
	* Makeconfig (COMBINE_INSMOD_RMMOD): New define.
	Requested by Erik Troan <ewt@redhat.com>.

Wed Sep 10 15:41:16 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* depmod/depmod.c (load_obj_file): Print error if fopen fails.
	(main): Continue processing on errors, so that we see all errors
	all at once.
	Reported by Gunther Mayer <gunther.mayer@braunschweig.netsurf.de>.

Tue Aug 26 11:14:36 -0400 1997  Telly Mavroidis  <mavroidi@ctc-fund.com>

	* util/sys_qm.c [__sparc__]: Fix typo.

Tue Jul 29 23:02:04 PDT 1997  Richard Henderson  <richard@gnu.ai.mit.edu>

	* include/module.h: Rename sizeof_type to tgt_sizeof_type.  Kill
	the struct duplication by defining a tgt_long.
	* insmod/insmod.c: Propogate sizeof_type change.

Tue Jul 29 22:23:16 PDT 1997  Jakub Jelinek  <jj@sunsite.mff.cuni.cz>

	* insmod/insmod.c (search_module_path): Add ipv6.

Tue Jul 29 22:23:16 PDT 1997  Jakub Jelinek  <jj@sunsite.mff.cuni.cz>

	Sparc64 support:
	* depmod/Makefile (DEFS): Pass in ARCH as well.
	* insmod/Makefile: Likewise.
	* obj/Makefile: Likewise.

	* insmod/insmod.c: Don't use native sizeof(type), use constants
	sizeof_type that correspond to the target.
	* include/module.h: Detect and cope with sparc64 target and
	sparc32 host.

	* include/obj.h: libc5 doesn't declare ELF64_ST_INFO either.

	* include/elf_sparc64.h: New file.
	* obj/obj_sparc64.c: New file.

Sat Jul 19 23:56:02 +0200 1997  Florian La Roche  <florian@knorke.saar.de>

	* depmod/misc.c (resolve_string): Kill assignment warning.
	* genksyms/genksyms.h: Extern outfile and debugfile properly.
	* man/modprobe.1: New file.

Wed Apr 16 15:27:17 +1000 1997  Paul Mackerras  <paulus@cs.anu.edu.au>

	* include/elf_ppc.h, obj/obj_ppc.c: New files.

Mon Apr 14 20:33:56 CDT 1997  Richard Henderson  <rth@tamu.edu>

	* depmod/depmod.c (print_deps_file): Fix initial array
	allocation size.

Sat Apr 12 19:23:25 CDT 1997  Richard Henderson  <rth@tamu.edu>

	* obj/obj_reloc.c (obj_relocate): If AXP_BROKEN_GAS, work
	around the old gas local symbol LITERAL reloc bug.

Mon Mar 31 23:03:12 EST 1997  David S. Miller  <davem@jenolan.rutgers.edu>

	* genksyms/lex.l (yylex) [case ST_TABLE_5]: Missing break.

Tue Mar 25 21:41:00 CST 1997  Richard Henderson  <rth@tamu.edu>

	* depmod/depmod.c (old_read_kernel_syms): Get symbols via
	get_kernel_syms syscall rather than /proc/ksyms.

	* depmod/modprobe.c (is_removable): Final arg to query_module
	is a size_t, not an int.

Sat Mar 15 19:34:07 PST 1997  H.J. Lu  <hjl@gnu.ai.mit.edu>

	* depmod/conf_file.c (release_all_sets): Set mod_set to NULL.
	(read_config_file): only call release_all_sets () once.
	call resolve_string () to expand string.

	* depmod/misc.c (resolve_string): New function that resolves
	string with `foo`.
	(what_command): New function that returns the known commands.
	(expand_command): New funtion that expands pattern or
	command. Only `uname -r` is supported this time.

	* depmod/misc.h: New prototype for resolve_string ().

	* depmod/modprobe.c (read_deps_file): call resolve_string () to
	expand string.

	* man/depmod.1: Fix typos. The syntax should be "path[TAG]=PATHNAME"
	means TAG is appended to PATHNAME.

Thu Mar 13 13:38:12 CST 1997  Richard Henderson  <rth@tamu.edu>

	* insmod/insmod.c (ncv_strcmp): New function that combines old
	m_strcmp and k_strcmp -- it's not a simple matter to always have
	the versioned symbol as the first or second argument.
	(main): Install ncv_strcmp instead.
	Reported by David Bourgin <dbourgin@wsc.com>.

	* insmod/insmod.c (get_kernel_version): Take an extra argument to
	return the full uname string.
	(old_get_module_version): Likewise.
	(new_get_module_version): Likewise.
	(main): Record and compare the full uname string, to handle
	suffixes like "-ISS" and "-SMP".

Sun Mar  9 23:52:18 CST 1997  Richard Henderson  <rth@tamu.edu>

	* genksyms/lex.l (yylex): Only use the file component of the input
	path when generating the output filename.
	Reported by Markus Dickebohm <m.dickebohm@uni-koeln.de>.

	* depmod/Makefile (clean): Remove modprobe and depmod.

Sun Mar  9 13:55:27 MET 1997  Eddie C. Dost  <ecd@skynet.be>

	* obj/obj_sparc.c (arch_apply_relocation): Handle lots
	more relocation types.

Tue Mar  4 08:12:58 CST 1997  Richard Henderson  <rth@tamu.edu>

	* Merge new depmod from Marcin Dalecki
	<dalecki@sub994.sub.uni-goettingen.de>.

	* insmod/insmod.c (get_kernel_version): Don't assert that the last
	number of the version be followed by a NUL.
	(old_get_module_version, new_get_module_version): Likewise.
	Reported by Keith Owens <kaos@ocs.com.au>

	* insmod/insmod.c: Merge character and string array patches
	from Jean Tourrilhes <jt@hplb.hpl.hp.com>.

	* kerneld/kerneld.c (kerneld_error): Add missing va_end; use vsnprintf.

Sun Feb  2 13:21:39 CST 1997  Richard Henderson  <rth@tamu.edu>

	* genksyms/lex.l (yylex): Update documentation on 2.0.x symbol table
	processing.  Generate phrases for asm blocks as well, as it is more
	efficient to do it in the lexer.
	* genksyms/parse.y (opt_asm_phrase, asm_declaration): Recognise 
	the ASM_PHRASE token rather than trying to work it out ourselves.

Sun Feb  2 07:44:00 CST 1997  Richard Henderson  <rth@tamu.edu>

	* man/kerneld.8: Format for section 8.
	* man/genksyms.8: Renamed from genksyms.1 to match how it formats.
	Reported by Martin von Loewis <martin@mira.isdn.cs.tu-berlin.de>.

Tue Jan 28 21:10:07 MET 1997  Janos Farkas  <chexum@shadow.banki.hu>


redhat-6.2-with-source/redhat-6.2/en_6.2/misc/src/anaconda/isys/modutils/depmod/depmod.c
/* Copyright 1997 Free Software Foundation, Inc.
   Contributed by Marcin Dalecki <dalecki@sub994.sub.uni-goettingen.de>

   This file is part of the Linux modutils.

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2 of the License, or (at your
   option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#ident "$Id: depmod.c,v 1.2 1999/09/11 15:43:57 msw Exp $"

#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <errno.h>
#include <elf.h>
#include ELF_MACHINE_H

#include "util.h"
#include "obj.h"
#include "misc.h"
#include "conf_file.h"

static int n_objs = 0;
static int max_objs = 0;
static struct obj_file **objs;
static char **names;

/*======================================================================*/

static int flag_show_error = 0;
static int flag_verbose = 0;

#ifndef NO_COMPAT
static int flag_new_syscalls = 0;
#endif

static struct new_module_symbol *ksyms;
static size_t nksyms;

/*======================================================================*/

static void
new_read_kernel_syms (void)
{
  struct new_module_symbol *syms, *s;
  size_t ret, bufsize, nsyms, j;

  /* Collect the kernel's symbols.  */

  syms = xmalloc (bufsize = 16 * 1024);
retry_kern_sym_load:
  if (query_module (NULL, QM_SYMBOLS, syms, bufsize, &ret))
    {
      if (errno == ENOSPC)
	{
	  syms = xrealloc (syms, bufsize = ret);
	  goto retry_kern_sym_load;
	}
      error ("kernel: QM_SYMBOLS: %m");
      exit (1);
    }
  nksyms = nsyms = ret;
  ksyms = syms;

  for (j = 0, s = syms; j < nsyms; ++j, ++s)
    s->name += (unsigned long) syms;
}

#ifndef NO_COMPAT

static void
old_read_kernel_syms (void)
{
  struct old_kernel_sym *ks, *k;
  struct new_module_symbol *s;
  int nks, nms, i;

  nks = get_kernel_syms(NULL);
  if (nks < 0)
    {
      error("get_kernel_syms: %m");
      return;
    }

  ks = k = xmalloc(nks * sizeof(*ks));
  if (get_kernel_syms(ks) != nks)
    {
      error("inconsistency with get_kernel_syms -- is someone else "
	    "playing with modules?");
      free(ks);
      return;
    }

  /* Collect the module information.  */

  while (k->name[0] != '#' || k->name[1])
    ++k;
  ++k;

  nksyms = nms = nks - (k - ks);
  ksyms = s = (nms ? xmalloc(nms * sizeof(*s)) : NULL);

  for (i = 0; i < nms; ++i, ++s, ++k)
    {
      s->name = (unsigned long)k->name;
      s->value = k->value;
    }
}

#endif

/*
 * Format the dependency list of a module into a simple makefile
 */
static void
print_deps_file (const char *depfile)
{
  FILE *fout = stdout;
  char **deps;
  int max_deps = 64;
  int o;

  deps = (char **) xmalloc (max_deps * sizeof (char *));

  if (depfile && !(fout = fopen (depfile, "w")))
    {
      lprintf ("can't open %s", depfile);
      exit (1);
    }

  /*
   * Loop through all modules we visited and construct the dependencies.
   */
  for (o = 0; o < n_objs; ++o)
    {
      int errs = 0;
      int n_deps = 0;
      int i;

      /*
       * Loop throught all undefined symbols of this object
       */
      for (i = 0; i < HASH_BUCKETS; ++i)
	{
	  struct obj_symbol *sym;
	  for (sym = objs[o]->symtab[i]; sym; sym = sym->next)
	    {
	      int j;
	      int k;

	      if (sym->secidx != SHN_UNDEF
		  || ELFW (ST_BIND) (sym->info) == STB_WEAK)
		continue;

	      /*
	       * Traverse all modules and search for one, where it may be
	       * defined.
	       */
	      for (j = 0; j < n_objs; ++j)
		{
		  struct obj_symbol *tmp;
		  if (j == o)
		    continue;

		  tmp = obj_find_symbol (objs[j], sym->name);
		  if (tmp && tmp->secidx != SHN_UNDEF)
		    break;
		}
	      if (j == n_objs)
		{
		  if (errs == 0)
		    lprintf ("%s: unresolved symbol(s)"
			     ,names[o]);
		  if (flag_show_error)
		    lprintf ("\t%s", sym->name);
		  errs++;
		  continue;
		}

	      /*
	       * Look if it's allready there
	       */
	      for (k = 0; k < n_deps; ++k)
		if (deps[k] == names[j])
		  break;
	      /*
	       * And add the dep info if neccessary.
	       */
	      if (k != n_deps)
		continue;

	      if (k == max_deps)
		{
		  max_deps <<= 1;
		  deps = (char **)
		    xrealloc (deps, max_deps * sizeof (char *));
		}
	      deps[k] = names[j];
	      ++n_deps;
	    }
	}

      if (errs == 0 && flag_verbose)
	printf ("%s\n", names[o]);
      fprintf (fout, "%s:", names[o]);

      /*
       * Print the dependencies for this module.
       */
#if 1
      while (n_deps--)
	fprintf (fout, " %s", deps[n_deps]);
#else
      for (i = 0; i < n_deps; ++i)
	fprintf (fout, " %s", deps[i]);
#endif
      fprintf (fout, "\n\n");
    }
  if (fout != stdout)
    fclose (fout);
  free (deps);
}

/*
 * Load an object file and resolve all the symbols contained therein.
 */
static int
load_obj_file (const char *module)
{
  FILE *fp;
  struct obj_file *file;
  struct obj_symbol *sym;
  int i;

  if (!(fp = fopen (module, "r")))
    {
      error("%s: open: %m", module);
      return 1;
    }

  file = obj_load (fp);
  fclose (fp);

  if (!file)
    {
      /* obj_load will already have printed an error message.  */
      /* there is a file in /lib/modules/ that cannot be loaded as
	 kernel module. as we could have something like "TRANS.TBL"
	 or other files in /lib/modules, we do not prevent writing
	 a "modules.dep" file. one or several wrong files is ok as
	 long as there are enough correct modules there... */
      return 0;
    }

  if (n_objs == max_objs)
    {
      max_objs <<= 2;
      objs = (struct obj_file **)
	xrealloc (objs, max_objs * sizeof (struct obj_file *));
      names = (char **) xrealloc (names, max_objs * sizeof (char *));
    }
  objs[n_objs] = file;
  names[n_objs] = xstrdup (module);

  /*
   * Hide the kernel symbols...
   */
  for (i = 0; i < nksyms; ++i)
    if ((sym = obj_find_symbol (file, (char *) ksyms[i].name)) != NULL)
      sym->secidx = SHN_HIRESERVE + 1;

  /*
   * Hide them they are used by insmod internally.
   */
  if ((sym = obj_find_symbol (file, "__this_module")) != NULL)
    sym->secidx = SHN_HIRESERVE + 1;
  if ((sym = obj_find_symbol (file, "mod_use_count_")) != NULL)
    sym->secidx = SHN_HIRESERVE + 1;
  ++n_objs;

  return 0;
}

/*
 * Print usage information ans exit.
 */
static void
usage (void)
{
  puts(
    "Usage: depmod [-e -s -v ] -a [FORCED_KERNEL_VER]\n"
    "       depmod [-e -s -v ] MODULE_1.o MODULE_2.o ...\n"
    "Create module-dependency information for modprobe.\n\n"
    "  -a, --all                  visit all modules\n"
    "  -d, --debug                run in debug mode\n"
    "  -e                         output unresolved symbols\n"
    "  -s, --system-log           use the system log for error reporting\n"
    "      --help                 display this help and exit\n"
    "  -v, --verbose              run in verbose mode\n"
    "  -V, --version              output version information and exit"
    );
}

int
main (int argc, char *argv[])
{
  int ret = 1;
  int flag_stdmode = 0;

  int opt_tag;

  if (argc == 1)
    {
      usage ();
      return 1;
    }

#ifndef NO_COMPAT
  flag_new_syscalls = !query_module (NULL, 0, NULL, 0, NULL);
  if (!flag_new_syscalls)
    old_read_kernel_syms ();
  else
#endif
    new_read_kernel_syms ();

  /*
   * Yes we are using getopts!
   */
  while (1)
    {
      static struct option long_opts[] =
      {
	{"all", 0, 0, 'a'},
	{"debug", 0, 0, 'd'},
	{"system-log", 0, 0, 's'},
	{"verbose", 0, 0, 'v'},
	{"version", 0, 0, 'V'},
	{"help", 0, 0, 'h'},
	{0, 0, 0, 0}		/* Table end tag */
      };
      int opt_ind = 0;

      opt_tag = getopt_long (argc, argv, "adesvV", long_opts,
			     &opt_ind);
      if (opt_tag == -1)
	break;

      switch (opt_tag)
	{
	case 'a':
	  flag_stdmode = 1;
	  break;

	case 'd':
	  flag_debug = 1;
	  break;

	case 'e':
	  flag_show_error = 1;
	  break;

	case 's':
	  setsyslog ("depmod");	/* use the syslog for reporting */
	  break;

	case 'v':
	  flag_verbose = 1;	/* give terse informations during run */
	  break;

	case '?':
	case 'h':
	  usage ();
	  exit (opt_tag == 'h' ? 0 : 1);
	  break;

	case 'V':
	  puts ("depmod (Linux modutils) " MODUTILS_VERSION);
	  if (argc != 2)
	    putchar ('\n');
	  break;

	default:
	  abort ();
	}
    }

  /*
   * Skip all automatically processed options.
   */
  argc -= optind;
  argv += optind;

  /*
   * argv now points to the first non-option argument
   * argc is the remaining argument count
   */

  n_objs = 0;
  max_objs = 64;
  objs = (struct obj_file **) xmalloc (max_objs * sizeof (struct obj_file *));
  names = (char **) xmalloc (max_objs * sizeof (char *));

  if (flag_stdmode)
    {
      struct mod_path *mods = NULL;
      struct mod_path *tmp;

      if (argc > 0)
	{
	  if (read_config_file (*argv))
	    {
	      lprintf ("%s does not exist", depfile);
	      return 1;
	    }
	}
      else
	{
	  /* Read the default config file and don't complain if there
	     isn't one.  */
	  read_config_file (NULL);
	}

      mods = find_matching_mods ("*", NULL, 1);
      ret = 0;
      while ((tmp = mods))
	{
	  ret |= load_obj_file (mods->path);
	  mods = mods->next;
	  free (tmp->path);
	  free (tmp);
	}
      while ((tmp = mods))
	{
	  mods = mods->next;
	  free (tmp->path);
	  free (tmp);
	}

      if (!ret)
	print_deps_file (depfile);
    }
  else
    {
      /*
       * Process all modules which are explicitly specified on the command line
       */

      for (ret = 0; argc > 0 && !ret; ++argv, --argc)
	ret = load_obj_file (*argv);
      if (!ret)
	print_deps_file (NULL);
    }

  return ret;
}
